README.md
# clojurebridgelondon-exercises-examples

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2016 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-09-27
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-09-27
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/clojurebridgelondon-exercises-examples/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clojurebridgelondon-exercises-examples/compare/0.1.0...0.1.1


test/clojurebridgelondon_exercises_examples/core_test.clj
(ns clojurebridgelondon-exercises-examples.core-test
  (:require [clojure.test :refer :all]
            [clojurebridgelondon-exercises-examples.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/clojurebridgelondon_exercises_examples/intro.clj
(ns clojurebridgelondon-exercises-examples.intro)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Introduction to Programming with Clojure
;;
;; - part of the older ClojureBridge curriculum


;; Some really basic concepts

;; Everything is a list in Clojure.  List stands for List Processing.  So what is a list and does it work

;; Really simplistic example - a shopping list

;; (shopping-in-tesco "eggs" "bacon" "cheese" "cerial" "milk") 

;; We have a list of things we want to buy.  The first item in the list defines how we are going to get these things

;; We need to write shopping-in-tesco function to make this work



;; EXERCISE: Basic arithmetic

;; - Take your height in feet and inches and convert it to inches using arithmetic in Clojure.
;; - Then convert that to centimeters. There are 2.54 centimeters in an inch.
;; - Lastly, ask two people near you for their height in centimeters. Find the average of your heights.

;; - Bonus: Convert that average back to feet and inches. The feet and the inches will be separate numbers. (quot x y) will give you the whole number part when dividing two numbers. (mod x y) will give you the remainder when dividing two numbers.

;; Define a persons height in feet and inches using a map and Clojure keywords for the keys
(def john {:feet 6 :inches 2})

;; Or to give this data more contenxt, an allow for further definitions of data for john, lets put the feet and inches data as the value for the height key
(def john {:height {:feet 6 :inches 2}})

john


(defn convert-feet-and-inches-to-feet
  "Convert feet and inches values to feet"
  [feet inches]
  (+ (* feet 12) inches))

(convert-feet-and-inches-to-feet (:feet (john :height))
                                 (:inches (john :height)))


(defn convert-inches-to-centimeters
  "Convert inches to centimeters"
  [inches]
  (* inches 2.54))

(convert-inches-to-centimeters 74)


;; This function could also be written using a local symbol for the conversion multiplier (a local symbol is like a local variable, but immutable of course because this is Clojure)
(defn convert-inches-to-centimeters-alternative
  "Convert inches to centimeters - using let to define a local symbol"
  [inches]
  (let [centimeter-multiplier 2.54]
    (* inches centimeter-multiplier)))

(convert-inches-to-centimeters-alternative 74)



;; The height for two other people
(def Anna  {:height {:feet 5 :inches 10}})
(def Bruce {:height {:feet 6 :inches 8}})


;; Calculate the average hight for all three people, John, Anna, Bruce

;; Steps to implement:
;; convert each person to centimeters < just start with Centimeters
;; calculate the average
;; convert back to feet & inches

;; clojure-bridge-exercises.core> (def heights [10 20 30])
;; #'clojure-bridge-exercises.core/heights
;; clojure-bridge-exercises.core> (count heights)
;; 3
;; clojure-bridge-exercises.core> (apply + heights)
;; 60


(defn average-height-in-centimeters
  "Given peoples height in centimeters, return the average height in centimeters"
  [heights]
  (/ (apply + heights) (count heights)))

(average-height-in-centimeters [180 220 140])

;; If the average is not a whole nuber, we get an interesting result
(average-height-in-centimeters [180 220 160])

;; Lets tweak the values and get a decimal number
(average-height-in-centimeters [180 220 160.0])


;; clojure-bridge-exercises.core> (quot 15 7)
;; 2
;; clojure-bridge-exercises.core> (mod 15 7)
;; 1

(defn convert-centimeters-to-feet-and-inches
  "Convert values in centimeters to feet and inches"
  [value]
  (let [feet (quot (/ value 2.54) 12)
        inches (mod (/ value 2.54) 12)]
    {:feet feet :inches inches}))

(defn average-height-in-feet-given-centimeters
  "Calculate the average hight of three people"
  [people]
  (convert-centimeters-to-feet-and-inches (average-height-in-centimeters people)))

(average-height-in-feet-given-centimeters [160 180 200])


src/clojurebridgelondon_exercises_examples/simple_values.clj
(ns clojurebridgelondon-exercises-examples.simple-values)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; EXERCISE 1: Basic arithmetic

;; How many minutes have elapsed since you arrived at the workshop today?
;; Convert this value from minutes to seconds.

;; eg. its been 2 hours and 5 minutes.  So that is 60 x 2 + 5 (+ (* 60 2) 5)

;; just putting in the values for 125 minutes since the workshop started
;; multiplying by 60 to get the seconds
(* 125 60)
;; => 7500



;;;;;;;;;;;;;;;;;;;;;;;;
;; To go a little further... we bind names to the values of hours and minutes
(def hours 2)
(def minutes 5)

;; we can then calculate the mintues in the hours so far
(* hours 60)
;; => 120

;; then work our the total time in minutes
(+ (* hours 60) minutes)
;; => 125

;; finally working out the total time in seconds
(* (+ (* hours 60) minutes) 60)
;; => 7500

;; formating to see the code more clearly
(* (+
    (* hours 60)
    minutes)
   60)
;; => 7500


;;;;;;;;;;;;;;;;;;;;;;;;
;; or writing this using the threading macro
(-> minutes
    (+ ,,, (* hours 60))
    (* ,,, 60))



;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; EXERCISE 2 [BONUS]: Minutes and seconds

;; Convert 1000 seconds to minutes and seconds.
;; The minutes and the seconds will be separate numbers.
;; (quot x y) will give you the whole number part of x divided by y.
;; (rem x y) will give you the remainder of x divided by y.

;; How many whole minutes in 1000 seconds?
(quot 1000 60)
;; => 16

;; How many seconds remain when you have taken all the whole minutes away from 1000 seconds
(rem 1000 60)
;; => 40

(str "There are " (quot 1000 60) " minutes and " (rem 1000 60) " seconds.")
;; => "There are 16 minutes and 40 seconds."

;; (/ 1000 60)





;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Simple values from old ClojureBridge curriculum

;;;;;;;;;;;;
;; simple values

"This is a simple string"
"Strings evaluate to themselves"
"Strings are immutable like other data structures in Clojure"
(str "Strings" "can" "be" "joined" "togehter" "using" "the" "str" "function")
(str "but str doesnt add" "spaces" "between strings")
(str "so add your own" " " "spaces")


;; Experimenting with joining strings
(def join-my-strings-with-spaces ["join" "us" "toether" "with" "spaces"])

;; (map str join-my-strings-with-spaces)

;; (apply str join-my-strings-with-spaces)

;; Using another function called interpose, we can place something between each string as we join them together
(apply str (interpose " " join-my-strings-with-spaces))

;; See https://clojuredocs.org/clojure.core/interpose


;; If you come from a Java world, you may want to use println to output a string.  Well you can

(println "This is my string, but caveat emptor, this is a side effect")

;; When you evaluate the above, you do not get a string in return... so what is going on?

;; Discuss evalation verses side effects here?


;; EXERCISE: Store the name of your hometown

"Ripon"

;; Write the name of your hometown as a string, and then assign that string to the name my-hometown.
;; EXERCISE: Make a function to format names

(def my-hometown "Ripon")

my-hometown

;; The str function can take any number of arguments, and it concatenates them together to make a string. Write a function called format-name that takes two arguments, first-name and last-name. This function should output the name like so: Last, First.

(defn format-name
  "Concatenate two strings together"
  [firstname lastname]
  (str firstname " " lastname))

(format-name "John" "Stevenson")

;; Adding in our interpose example from before...

(defn format-name
  "Concatenate two strings together"
  [firstname lastname]
  (interpose " " (str firstname lastname)))

(format-name "John" "Stevenson")

;; Thats not quite right.  If we change the arguments to a vector then we can

(defn format-name
  "Concatenate two strings together"
  [strings-to-join]
  (apply str (interpose " " strings-to-join)))

(format-name ["John" "Stevenson"])



src/clojurebridgelondon_exercises_examples/data_structures.clj
(ns clojurebridgelondon-exercises-examples.data-structures)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; EXERCISE 1: See turtle names

;; Uses turtle walk.clj from the welcometoclojurebridge project, not this project.



;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; EXERCISE 2: Make a vector

;; Go to myproject’s core.clj and start InstaREPL
;; Make a vector of the high temperatures for the next 7 days in the town where you live.
;; Then use the nth function to get the high temperature for next Tuesday.

;; A simple vector just holding the temperatures in celceus for each day of the week
[23 24 25 29 32 18 17]

;; If we want the temperature of a particular day, we can use one of the sequence fuctions, eg. nth
;; The nth function uses the vectors index to find the value, the vectors index starts at zero.

;; Assuming this vector starts with Monday (position 0), then Tuesday will be at position 1

(nth [23 24 25 29 32 18 17] 1)
;; => 24

;; If we were going to use thes temperature readings multiple times, we could give them a names

(def london-temperatures [23 24 25 29 32 18 17])

(nth london-temperatures 2)
;; => 25

(first london-temperatures)
;; => 23

(last london-temperatures)
;; => 17

;;;;;;;;;;;;;;;;;;;;;;;;
;; Additional: what if we paired the temperatures with their days
;; This would make it easier to reason about the values

;; We could use strings for the keys, for each day of the week
{"Monday" 23 "Tuesday" 24 "Wednesday" 25 "Thursday" 29 "Friday" 32 "Saturday" 18 "Sunday" 17}

;; Lets give this a name (symbol) so we can use it easily in other places
(def temperatures-strings {"Monday" 23 "Tuesday" 24 "Wednesday" 25 "Thursday" 29 "Friday" 32 "Saturday" 18 "Sunday" 17})

;; The get function will return the value assosicate with the specified key
(get temperatures-strings "Monday")
;; => 23

;; We can also write this in shorthand, using the fact that maps can get values given a key
(temperatures-strings "Monday")

;; We can also use keywords as the keys.
{:Monday 23 :Tuesday 24 :Wednesday 25 :Thursday 29 :Friday 32 :Saturday 18 :Sunday 17}

;; And give this a name too
(def temperatures-keywords {:Monday 23 :Tuesday 24 :Wednesday 25 :Thursday 29 :Friday 32 :Saturday 18 :Sunday 17})

;; The get function works the same
(get temperatures-keywords :Monday)
;; => 23

;; As before we can use just the map and keyword
(temperatures-keywords :Monday)
;; => 23

;; As we are using keywords for keys, we can also do a little short-cut
(:Monday temperatures-keywords)
;; => 23


;;;;;;;;;;;;;;;;;;;;;;;;
;; Alternative approach - using days of the week as a lookup map

;; Make a vector of the high temperatues for the next 7 days in the town where you live. Then use the nth function to get the high temperature for next Tuesday.

(def weather-temperatures-forecast [5 7 13 12 11 12 9])

(nth weather-temperatures-forecast 1)

;; We could make this clearer with a nice little map to represent days of the week

(def days-of-the-week {:monday 0
                                :tuesday 1
                                :wednesday 2
                                :thursday 3
                                :friday 4
                                :saturday 5
                                :sunday 6})

(nth weather-temperatures-forecast (days-of-the-week :tuesday))

;; or even fancier with the threading macro

(->
 weather-temperatures-forecast
 (nth (days-of-the-week :tuesday)))


;; Making a longer range forcast by conjoin (joining) more temperatures to our forcast data structure

(conj weather-temperatures-forecast 13 15 19 20 22 21 24)

;; So what does the forcast data structure look like now?

weather-temperatures-forecast

;; Should we wish to keep this longer forcast, we can assign a name to it (makeing a clojure symbol)

(def long-range-weather-forecast
  (conj weather-temperatures-forecast 13 15 19 20 22 21 24))

;; Now we can access the full long range forecast
long-range-weather-forecast

;; Although underlying data structures cannont be changed in Clojure (they are immutable),
;; you can change names and point them to new data structures

;; So if we want it to look like we have updated our original forcast data, then we can point the original
;; name to the new data structure we are creating
(def weather-temperatures-forecast
  (conj weather-temperatures-forecast 13 15 19 20 22 21 24))

;; When we re-evaluate the weather-temperature-forecast it has all the new values
weather-temperatures-forecast





;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; EXERCISE 3: See turtles states

;; Go to walk.clj file, not this project


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; EXERCISE 4: Modeling Yourself

;; Use the myproject’s core.clj and InstaREPL
;; Make a map representing yourself
;; Make sure it contains your first name and last name
;; Then, add your hometown to the map using assoc or merge


;; You could create the simplest possible
;; Here the key is my first name and value is my last name
{"John" "Stevenson"}

;; To make the map clearer, you could use keys that define some context for the values
{"name" "John Stevenson"}
{"firstname" "John" "lastname" "Stevenson"}


;; To add a key / value pair to a map you can use assoc or merge

;; (assoc map keys value)

(assoc {"firstname" "John" "lastname" "Stevenson"} "hometown" "Harrogate")
;; => {"firstname" "John", "lastname" "Stevenson", "hometown" "Harrogate"}

;;(merge map map)

(merge {"firstname" "John" "lastname" "Stevenson"} {"hometown" "Harrogate"})
;; => {"firstname" "John", "lastname" "Stevenson", "hometown" "Harrogate"}

;; NOTE: You may have noticed the results have comma seperated pairs.  Commas have no meaning in Clojure, they are just there to help the developer.

;;;;;;;;;;;;;;;;;;;;;;;;
;; Previously we have used strings for the keys, we can also use Clojure keywords for the keys.
;; Keywords point to themselves (as apposed to bindins / names that point to another value).
;; Keywords can be used to make finding things in maps really easy.

;; We could have a simple name key and fullname as a value
{:name "John Stevenson"}

;; Or split the name into first and last keys and values
{:firstname "John" :lastname "Stevenson"}

;; Association still works the same
(assoc {:firstname "John" :lastname "Stevenson"} :hometown "Harrogate")
;; => {:firstname "John", :lastname "Stevenson", :hometown "Harrogate"}

(merge {:firstname "John" :lastname "Stevenson"} {:hometown "Harrogate"})


;;;;;;;;;;;;;;;;;;;;;;;;
;; Taking things further

;; You can go further and create a nested map
{:name {:first "John" :last "Stevenson"}}

;; You can then add other keys at the top level as before
(assoc {:name {:first "John" :last "Stevenson"}} :hometown "Harrogate")
;; => {:name {:first "John", :last "Stevenson"}
;;     :hometown "Harrogate"}

(merge {:name {:first "John" :last "Stevenson"}} {:hometown "Harrogate"})
;; => {:name {:first "John", :last "Stevenson"}
;;     :hometown "Harrogate"}


;; The assoc and merge add values to a map.
;; You can also update the values in a map using the update and update-in functions

(def johns-details {:name {:first "John" :last "Stevenson"}
                    :hometown "Harrogate"
                    :age 24})

;; The update function takes a map, key and function as arguemnts.
;; The function takes the existing value as an argument.
;; In this example we increment my age by 1

(update johns-details :age inc)
;; => {:name {:first "John", :last "Stevenson"}
;;     :hometown "Harrogate"
;;     :age 25}



;; EXERCISE 5 [BONUS]: Modeling your classmates

;; First, take the map you made about yourself in previous exercise.
;; Then, create a vector of maps containing the first name, last name and hometown of two or three other classmates around you.
;; Lastly, add your map to their information using conj.

{:firstname "John" :lastname "Stevenson"}

[{:firstname "John"  :lastname "Stevenson"}
 {:firstname "Betty" :lastname "Smith"}
 {:firstname "Jane"  :lastname "Aire"}]

(def people [{:firstname "John"  :lastname "Stevenson"}
             {:firstname "Betty" :lastname "Smith"}
             {:firstname "Jane"  :lastname "Aire"}])


;; To get values from such a data structure, you can get individual maps
(first people)
;; => {:firstname "John", :lastname "Stevenson"}

(get (first people) :firstname)
;; => "John"

(conj people {:firstname "Susan" :lastname "Davros"})
;; => [{:firstname "John", :lastname "Stevenson"}
;;     {:firstname "Betty", :lastname "Smith"}
;;     {:firstname "Jane", :lastname "Aire"}
;;     {:firstname "Susan", :lastname "Davros"}]


;;;;;;;;;;;;;;;;;;;;;;;;
;; Additional examples

;; To get all the first names in the data structure we use the get function,
;; however, as we are working over a collection, we wrap the get function
;; in an anonymous function.  Then we map this anonymous function over each
;; element of the collection and collect the results in a list that we return
;; as the result.
;; The anonymous function uses the % to indicate where each element is placed
;; in turn, just before the get function is evaluated.

(map #(get % :firstname) people)
;; => ("John" "Betty" "Jane")

;; and similarly for all the last names
(map #(get % :lastname) people)
;; => ("Stevenson" "Smith" "Aire")


src/clojurebridgelondon_exercises_examples/functions.clj
(ns clojurebridgelondon-exercises-examples.functions)


;; EXERCISE 1: Move turtles using function

;; Uses turtle walk.clj from the welcometoclojurebridge project, not this project.

;; EXERCISE 2: Move turtles using function with parameters

;; Uses turtle walk.clj from the welcometoclojurebridge project, not this project.



;; predicates

;; =, not=
;; >, <, >=, <=
;; true?, false?, empty?, nil?, vector?, map?


;; EXERCISE 3 [BONUS]: Find the average

;; Create a function called average that takes a vector of maps.
;; Use [{:angle 30} {:angle 90} {:angle 50}] as input.
;; Calculate average value of :angle.
;; Hint: You will need to use the functions map, reduce and count.


;; Lets start with a simpler example
;; We create a function that takes three angles
;; We add the angles together and divide by 3
(defn average-simple
  [angle1 angle2 angle3]
  (/ (+ angle1 angle2 angle3) 3))

(average-simple 30 90 50)


;; In the exercise itself, our values are contained within a collection, in this case a map.
;; So to solve this exercise, at some point we need to extract our values.

;; Previously we worked with values in a collection (a map) and used the get function
;; We also used the feature of keywords to help us get values from maps.
;; We can use one of the following to extract values from our collection of angles

(get {:angle 30} :angle)
;; => 30

(:angle {:angle 30})
;; => 30

;; When we have several maps in a vector, we need to iterate over each map
;; and extract each of the values.
;; We can use the map function to extract all the values of each map in the vector
;; however the get function we nomally would use isnt sufficient because the
;; get function requires two arguments, the map & the key

;; Neither of these expressions are correct because they call the get function incorrectly.
(map get [{:angle 30} {:angle 90} {:angle 50}] :angle)
(map get :angle [{:angle 30} {:angle 90} {:angle 50}])

;; To solve this we can use a fuction called partial.
;; The partial function allows you to call another function, but with a missing argument

;; In a simple example we add 90 to every number in the collection
(map (partial + 90) [0 30 60 90])
;; => (90 120 150 180)


;; For the exercise we use partial with the :angle keyword, which is passed each map in turn
;; as a second argument by the map function.
(map (partial :angle) [{:angle 30} {:angle 90} {:angle 50}])
;; => (30 90 50)

;; Now we can get all the values returned in a collection, we need to add them.
;; If we want to take all the values in a collection and get a single value
;; then you can use the reduce function.
;; Here we reduce these values by adding them together
(reduce + [30 90 50])


;; Lets give a name to our angles data, to make it easy to work with
(def angles [{:angle 30} {:angle 90} {:angle 50}])

;; Now we can check our map expression
(map (partial :angle) angles)
;; => (30 90 50)

;; We want the total value of all the angles, so we reduce them with the + function into one value
(reduce + (map (partial :angle) angles))


;; Now we need the average, so we divide the total value by the number of angle values
;; In this example, the count function returns how many maps are in the vector (there is one angle value per map)
(/ (reduce + (map (partial :angle) angles)) (count angles))
;; => 170/3

;; Now we can put this all in a function definition, so we can call it multiple times (if needed)
(defn average [angles]
  (/ (reduce + (map (partial :angle) angles)) (count angles)))

;; Now we call the average function with the arguments from the exercise
(average [{:angle 30} {:angle 90} {:angle 50}])

;; As we use the count function to help us find the average, we can call average with any number of maps in our vector
(average [{:angle 30} {:angle 90} {:angle 50} {:angle 60}])
;; => 115/2

(average [{:angle 30} {:angle 90} {:angle 50} {:angle 60} {:angle 80}])
;; => 62



;;;;;;;;;;;;;;;;;;;;;;;;
;; Alternative - anonymous function

;; Instead of using the partial function, you can use an anonymous function.
;; In this case we are using the short-hand syntax for an anonymous function.
(map #(get % :angle) angles)
;; => (30 90 50)

(reduce + (map #(get % :angle) angles))
;; => 170

(/ (reduce + (map #(get % :angle) angles)) (count angles))
;; => 170/3


;;;;;;;;;;;;;;;;;;;;;;;;
;; Let - local names / bindings

;; Here we use the let function to bind the name number-of-angles
;; to the number of angle maps passed in as the argument.
;; We can then use number-of-angles as the value to divide the total,
;; giving us the average value.
;; Local bindings are only within the scope of the let expression,
;; so calling number-of-angles after the closing perentesis of the let will cause an error.
(defn average [angles]
  (let [number-of-angles (count angles)]
    (/ (reduce + (map (partial :angle) angles)) number-of-angles)))

(average [{:angle 30} {:angle 90} {:angle 50}])
;; => 170/3




;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Functions - older ClojureBridge curiculum


;; EXERCISE: Find per-person share of bill among a group

;; Create a new function called share-per-person.

;; Modify our total-with-tip function, and call the new function share-per-person, that additionally takes in as an argument the number of people in the group for a bill. Have the function return the average bill amount per person.

(defn share-per-person
  "Share the cost of a bill evenly for everyone at the meal"
  [meal-cost people]
  (/ meal-cost people))


(defn total-with-tip-per-person
  "Given subtotal, return total after tax and tip per person (hard-coded)."
  [subtotal tip-percent]
  (share-per-person (* 1.08 subtotal (+ 1 tip-percent)) 2))

(total-with-tip-per-person 12.50 0.18)
(total-with-tip-per-person 12.50 0)



;; EXERCISE: Find the average

;; Create a function called average that takes a vector of bill amounts and returns the average of those amounts.

;; Hint: You will need to use the functions reduce and count.

;; Define a collection of the orders for a group of people at a resteraunt
(def dine-in-orders [12.50 20 21 16 18.40])

;; Define a collection of orders for a take-away for a party
(def take-out-orders [6.00 6.00 7.95 6.25])

;; Use the previously defined function, total-bill, to add tax to each of the order costs in the collection
(map total-bill dine-in-orders)  ;=> [13.5 21.6 22.68 17.28 19.872]
(map total-bill take-out-orders) ;=> [6.48 6.48 8.586 6.75]


;; Lets look at map and reduce functions

(reduce + dine-in-orders)

;; So we can easily get the total cost of the orders, the average per person is simply the total divided by the number of orders

;; Lets get the number of orders from the collection
(count dine-in-orders)

;; So joining these two together we can get the average
(/ (reduce + dine-in-orders) (count dine-in-orders))

;; Lets put this into a function so we have a name for this behaviour which we can call

(defn total-bill-per-person
  "Given subtotal of bill, return total after tax."
  [individual-order-costs]
  (/ (reduce + dine-in-orders) (count dine-in-orders)))

(total-bill-per-person dine-in-orders)


;; But wait, we have forgotten to include sales tax

(defn total-bill-per-person-with-tax
  "Given subtotal of bill, return total after tax."
  [individual-order-costs]
  (/ (reduce + (map total-bill dine-in-orders)) (count dine-in-orders)))

(total-bill-per-person dine-in-orders)

;; Oh no, we forgot the tip as well....

(defn total-bill-per-person-with-tax-and-tip
  "Given subtotal of bill, return total after tax."
  [individual-order-costs]
  (/ (reduce + (map total-with-tip dine-in-orders )) (count dine-in-orders)))

(total-bill-per-person dine-in-orders)

;;; This is not quite right, as we need to pass in the percentage of tip.... not sure how to do that, perhaps a partial function?

;; We can use the function apply instead of reduce.  These functions can seem similar in concept, so lets see what we can do with apply to help you understand the difference


;; Lets get the total cost of the dine-in orders
(apply + dine-in-orders)

;; Now if we divide by the number of orders in the collction, we can get the average
(/ (apply + dine-in-orders) (count dine-in-orders) )

;; This all looks the same, so lets look at the Clojure docs to try understand the difference


;; > Note: Clojure is strict when it comes to functions as arguments.  This means that a function pased as an argument to another function is evaluated before it is passed as the argument.  So a function always recieves values as arguments, because a function always evaluated to a value (even if that value is nil).




;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Functions from old ClojureBridge content
;;
;; - includes some more advanced examples

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; More functions


;; Note for existing devs:  Clojure does not use = for assignment, so you dont need hacks like == for assignement.
;; Clojure keeps it simple


;; Naming conventions: Predicates?
;; When a function is asking a question where the answer is either true or false, the naming convention is to add a ? to the end of the function name.  for example
;; (true? (= 1 1))
;; (false? (= 1 2))

;; (vector? [1 2 3])



;; EXERCISE: Modeling Yourself

;; Make a map representing yourself. Make sure it contains your first name and last name. Then, add your hometown to the map using assoc or merge.

;; Hint: First, create a function that returns the name when given a single person's map.  

(def john
  {:firstname "john" :lastname "stevenson"})

(merge john {:hometomwn "Ripon"})
(assoc john :hometomwn "Ripon")


;; EXERCISE 2: Get the names of people

;; Create a function called get-names that takes a vector of maps of people and returns a vector of their names.

;; Here is an example of how it should work:

(get-names [{:first "Margaret" :last "Atwood"}
            {:first "Doris" :last "Lessing"}
            {:first "Ursula" :last "Le Guin"}
            {:first "Alice" :last "Munro"}])

;=> ["Margaret Atwood" "Doris Lessing" "Ursula Le Guin" "Alice Munro"]


(defn get-names [person]
  (let [name (:name person)
        age  (:age person)]
    (str name)))


(macroexpand '(let [name (:name person)
                    age (:age person)]
                (str name)))

;; (macroexpand let*)

;; A shorthand version of get-name function, using the keyword as the get function
(defn get-names-shorthand [person]
  (:name person))

(get-names {:name "john stevenson" :age 21})




;; EXERCISE: Modeling your classmates

;; First, take the map you made about yourself.

;; Then, create a vector of maps containing the first name, last name and hometown of two or three other classmates around you.

;; Lastly, add your map to their information using conj.

;; Use the get-names function from the previous Exercise to output a list of the names.





;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Predicates, Sequence comprehension & laziness

;; Predicates
;; A predicate is function that will return true or false.
;; The naming convention in Clojure is to add a ? to the end of the name.

(even? 7)
(even? 8)

;; Range generates an infinate sequence, however using the take function
;; a specific sequence range is returned.
(take 1000 (range))

;; return only the even numbers
(def evens
  (filter even? (take 1000 (range))))


;; Using an anonymous function with the filter function to only return values
;; where the value of dividing a value by 3 has an even remainder
;; (filter (fn [x] (even? (mod x 3))) '(1 2 3 4 5 6))

(def evens
  (filter (fn [x] (even? (mod x 3))) (filter even? (take 1000 (range)))))

evens

;; map, take, drop, filter, rest :: sequence->sequence
;; reduce, first  :: sequence->single
(def evens
  (->>
   (range)
   (take 1000)
   (filter even?)
   (filter (fn [x] (even? (mod x 3))))))


;;;;;;;;;;;;;;;;;;;
;; Threading Macros

;; Using the Threading macros provides a simple way to chain functions together
;; The reader macro for comments can be used to only run specific functions in the chain.
;; In this Thread first example, the value of the previous function is the first
;; argument to the next function
;; Uncomment each line and evaluate to see the value returned by each expression
(->
  {}
  #_(assoc ,,, :x {:x1 8})
  #_(update-in ,,, [:x :x1] inc)
  #_(assoc ,,, :y 99 :z 199999)
  #_(dissoc ,,, :z))


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Understanding Clojure macros

;; Macros provide a way to extend the Clojure syntax and are typically used
;; to maintain simplicity in the language.
;; To look under the covers of Clojure macros you can use the macroexpand function
;; (clojure.walk/macroexpand-all ')

(clojure.walk/macroexpand-all '(defn expand-me [args] (str "behaviour of function")))

;; expands to
;; (def expand-me (fn* ([args] (str "behaviour of function"))))
